File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int maxLength(vector<string>& arr) {
        vector<bitset<26>> dp = { bitset<26>() };
        int res = 0; // will store number of unique chars in resultant string
        for (auto& s : arr) {
            bitset<26> a; // used to track unique chars
            for (char c : s) {
                a.set(c - 'a');
            }
            int n = a.count();
            if (n < s.size()) {
                continue; // duplicate chars in current string
            }

            for (int i = dp.size() - 1; i >= 0; i--) {
                bitset<26> c = dp[i];
                if ((c & a).any()) {
                    continue; // if 1 or more char common
                }
                dp.push_back(c | a); // valid concatenation
                res = max(res, (int)c.count() + n);
            }
        }
        return res;
    }
};

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    public int maxLength(List<String> arr) {
        List<Integer> dp = new ArrayList<>();
        dp.add(0);
        int result = 0;
        for (String str : arr) {
            int a = 0, duplicate = 0;
            for (char c : str.toCharArray()) {
                duplicate |= a & (1 << (c - 'a'));
                a |= 1 << (c - 'a');
            }
            if (duplicate > 0) {
                continue;
            }
            for (int i = dp.size() - 1; i >= 0; i--) {
                if ((dp.get(i) & a) > 0) {
                    continue;
                }
                dp.add(dp.get(i) | a);
                result = Math.max(result, Integer.bitCount(dp.get(i) | a));
            }
            
        }
        return result;
    }
}